### DISCUSSION TITLE
Is there a recommended way to mock fetch?

### DISCUSSION THREAD

**Discussion by dusty-phillips** [oven-sh/bun#11169] [Q&A]
I've got a (bun + hono) api that interacts with several different REST APIs. Some of those APIs are easy to test; I control the code that calls fetch and can mock it sanely. But others are using SDKs that call fetch on my behalf (sometimes the global fetch, other times node-fetch).

Is there a good way to reliably mock fetch in these third-party projects? I can mock the individual SDK callsites if I have to, but it would be easier to mock fetch. I've tried several incantations of `mock.module` and even setting `global.fetch = mock()`, but haven't had much luck.

**Comment by dusty-phillips** [comment]
I wouldn't call it "recommended" but this seems to work:

```ts
import { afterEach, mock } from "bun:test"

export const mockFetch = mock()

global.fetch = mockFetch
mock.module("node-fetch", () => mockFetch)

afterEach(() => {
  mockFetch.mockReset()
})
```

Then in a test file:

```ts
import { mockFetch } from "./mock-fetch-preload"

test("Fetch gets mocked again", async () => {
  mockFetch.mockReturnValueOnce(Promise.resolve(new Response("hello world")))
  const response = await <some function that calls fetch>
  expect(mockFetch).toHaveBeenCalledTimes(1)
  expect(mockFetch).toHaveBeenCalledWith(...)
  //etc
})
```

It needs to be run with `bun test --preload ./path/to/module.ts` and seems to be mocking reliably... so far.

**Reply by ImBIOS** [reply]
```
# Unhandled error between tests
-------------------------------
1 | (function (entry, fetcher)
    ^
SyntaxError: Missing 'default' export in module 'node-fetch'.
-------------------------------


 0 pass
 1 fail
 1 error
Ran 1 tests across 1 files. [93.00ms]
```

**Comment by vitch** [comment]
[This](https://x.com/jarredsumner/status/1578833646089183232) gives hope there is something better but I guess not built in...

I did just come across [bun-bagel](https://github.com/DRFR0ST/bun-bagel) which sounds promising and I'm going to try out.

**Reply by nicholasoxford** [reply]
I would recommend [bun-bagel](https://github.com/DRFR0ST/bun-bagel) to anyone coming across this thread. Within minutes I had fetch mocked the way I imagined

> [This](https://x.com/jarredsumner/status/1578833646089183232) gives hope there is something better but I guess not built in...
> 
> I did just come across [bun-bagel](https://github.com/DRFR0ST/bun-bagel) which sounds promising and I'm going to try out.

**Reply by owenizedd** [reply]
The problem with bun-bagel is that it can't mock the return header as you want. it only support body mocking.

**Reply by DRFR0ST** [reply]
@owenizedd working on it! 😄  https://github.com/DRFR0ST/bun-bagel/pull/7

**Reply by rafaell-lycan** [reply]
@DRFR0ST is this gonna be a better (simple) version of `nock`? 💯 

I have solved my issues with `undici` while I couldn't solve them with `nock`.

Nit tip: To mock response errors isn't working properly, or perhaps the guides are outdated.

@Jarred-Sumner is there an easy way to mock `fetch` in bun? Perhaps a `mock.fetch` will be cooked soon?

thx

**Reply by DRFR0ST** [reply]
@rafaell-lycan initially I've built the library for my personal use, but decided to make it public. Not sure yet where it will go, but I'm semi-actively improving it upon demand, at least as long there is no official solution.

Thanks, I'll look into the response error issue. 

Contributions are very welcome and appreciated! 🙏

**Comment by kravetsone** [comment]
> I've got a (bun + hono) api that interacts with several different REST APIs. Some of those APIs are easy to test; I control the code that calls fetch and can mock it sanely. But others are using SDKs that call fetch on my behalf (sometimes the global fetch, other times node-fetch).
> 
> Is there a good way to reliably mock fetch in these third-party projects? I can mock the individual SDK callsites if I have to, but it would be easier to mock fetch. I've tried several incantations of `mock.module` and even setting `global.fetch = mock()`, but haven't had much luck.

Msw

**Comment by wellcaffeinated** [comment]
For future googlers. I found bun-bagel to be buggy and feature incomplete. This worked better for me:
https://github.com/aryzing/bun-mock-fetch

**Comment by lilienblum** [comment]
**UPDATE:** published package https://github.com/lilienblum/bun-fetch-mock

I usually use [spyOn](https://bun.sh/guides/test/spy-on):

```ts
const mockGetExample: typeof fetch = async (url, opts) => {
	expect(url).toBe('https://example.com')
	expect(opts?.method).toBe('GET')

	return new Response(`{"success":true}`)
}

const spyFetch = spyOn(globalThis, 'fetch')

beforeEach(() => {
	spyFetch.mockReset()
})

afterAll(() => {
	spyFetch.mockRestore()
})

test('fetches example.com', () => {
	spyFetch.mockImplementation(mockGetExample)
	...
```

**Comment by ScottKayeNeo** [comment]
For anyone coming back to this thread, I solved this like this (thanks @dn-l):

```ts
// Your code
export async function doFetchThing() {
  return fetch('http://localhost:1234').then((r) => r.json());
}
```

```ts
import { describe, expect, spyOn, test } from 'bun:test';

import { doFetchThing } from './wherever';

describe('...', () => {
  test('...', async () => {
    spyOn(global, 'fetch').mockResolvedValueOnce(
      Response.json({
        hey: 123,
      }),
    );

    const result = await doFetchThing();

    expect(result).toEqual({ hey: 123 });
  });
});

```

`global` + `spyOn` + `Response.json()` makes this really nice!

**Comment by lilienblum** [comment]
Wrapped as a package https://github.com/lilienblum/bun-fetch-mock
